cluster-autoscaler/proposals/provisioning-request.md

```diff
@@ -0,0 +1,293 @@
+}
+
+type PodSet struct {
+	// Template representing pods that will consume this reservation.
+	// Requirements for resources (CPU, RAM, GPUs, TPUs, storage) are
+	// necessary (must be non-zero). Users need to make sure that the
+	// tolerations and label selectors are consistent between this template
+	// and actual pods consuming the Provisioning Request.
+	Template apiv1.PodTemplateSpec `json:"template" protobuf:"bytes,1,name=template"`
+	// Count contains the number of pods that will be created with a given
+	// template.
+	// +kubebuilder:validation:Minimum=1
+	Count int32 `json:"count" protobuf:"bytes,2,name=count"`
+}
+
+// ProvisioningRequestStatus represents the status of the resource reservation.
+type ProvisioningRequestStatus struct {
+	// Conditions represent the observations of a Provisioning Request's
+	// current state. Those will contain information whether the capacity
+    // was found/created or if there were any issues. The condition types
+    // may differ between different provisioning classes.
+	//
+	// +optional
+	Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,1,rep,name=conditions"`
+
+	// AdditionalStatus contains all other status values custom provisioning classes may require.
+	//
+	// +optional
+	AdditionalStatus map[string]string `json:"additionalStatus" protobuf:"bytes,2,name=additionalStatus"`
+}
+```
+
+### Provisioning Classes
+
+#### GenericCheckCapacity class
+
+The `GenericCheckCapacity` is one-off check to verify that the in the cluster
+there is enough capacity to provision given set of pods.
+
+Note: If two of such objects are created around the same time, CA will consider
+them independently and place no guards for the capacity.
+Also the capacity is not reserved in any manner so it may be scaled-down.
+
+#### GenericAtomicScaleUp class
+
+The `GenericAtomicScaleUp` aims to provision the resources required for the
+specified pods in an atomic way. The proposed logic is to:
+1. Try to provision required VMs in one loop.
+2. If it failed, remove the partially provisioned VMs and back-off.
+3. Stop the back-off after a given duration (optional), which would be passed
+   via `ValidUntilSeconds` field and value containing seconds for which we should 
+   retry (measured since creation fo the CR).
+
+Note: that the VMs created in this mode are subject to the scale-down logic.
+So the duration during which users need to create the Pods is equal to the
+value of `--scale-down-unneeded-time` flag.
+
+### Adding pods that consume given ProvisioningRequest
+
+To avoid generating double scale-ups and exclude pods that are meant to consume
+given capacity CA should be able to differentiate those from all other pods.  
+To do so users need to specify the following pod annotation (it is not required
+in ProvReq’s template, though it can be specified):
+
+```yaml
+annotations:
+    "cluster-autoscaler.kubernetes.io/consume-provisioning-request": "provreq-name"
```

Comment:
since we have to update pods to refer back to provisionrequest anyway, is there a way to simplify/trim down/deduplicate the info embedded in provisionrequest objects and pull the pod spec info either from pod or podtemplate objects that associate themselves with the provisionrequest?

For example,

provisioningrequest `request-1` just contains the info about the existence/count of pods in podsets:
```
podSets: [
{name:"foo",count:100},
{name:"bar",count:100},
{name:"baz",count:100},
]
```

and cluster autoscaler would wait until it sees pods (or podtemplates, if pre-provisioning was desired) identifying as members of `request-1` and has podspecs for all podsets:
```
"cluster-autoscaler.kubernetes.io/provisioning-request": "request-1"
"cluster-autoscaler.kubernetes.io/podset": "foo"
```

Reply:
Adding support for podTemplate is definitely something we would like to do, but referencing the pod (or workload definitions) themselves seems to generate more complexity that it allows us to hide by this CRD. 

The idea with the CRD is that users do not need to create the pods up front and create them once there is capacity in cluster.

Reply:
Embedding more and more extensive schemas in CRDs doesn't seem like it will work well long-term (due to schema skew, validation limitations, and CRD / CR size). A PodTemplate object would also allow recording the podspec without creating pods up front

Reply:
we will proceed with referencing the PodTempalte